sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala

```diff
@@ -515,28 +515,66 @@ case class AlterTableRenamePartitionCommand(
  */
 case class AlterTableDropPartitionCommand(
     tableName: TableIdentifier,
-    specs: Seq[TablePartitionSpec],
+    partitions: Seq[(TablePartitionSpec, Seq[Expression])],
     ifExists: Boolean,
     purge: Boolean,
     retainData: Boolean)
-  extends RunnableCommand {
+  extends RunnableCommand with PredicateHelper {
 
   override def run(sparkSession: SparkSession): Seq[Row] = {
     val catalog = sparkSession.sessionState.catalog
     val table = catalog.getTableMetadata(tableName)
+    val resolver = sparkSession.sessionState.conf.resolver
     DDLUtils.verifyAlterTableType(catalog, table, isView = false)
     DDLUtils.verifyPartitionProviderIsHive(sparkSession, table, "ALTER TABLE DROP PARTITION")
 
-    val normalizedSpecs = specs.map { spec =>
-      PartitioningUtils.normalizePartitionSpec(
-        spec,
-        table.partitionColumnNames,
-        table.identifier.quotedString,
-        sparkSession.sessionState.conf.resolver)
+    val toDrop = partitions.flatMap { partition =>
+      val normalizedSpecs = PartitioningUtils.normalizePartitionSpec(
+            partition._1,
+            table.partitionColumnNames,
+            table.identifier.quotedString,
+            sparkSession.sessionState.conf.resolver)
+
+      val partitionSet = {
+        if (partition._2.nonEmpty) {
+          val parts = partition._2.map { expr =>
+            val (attrName, value) = expr match {
+              case BinaryComparison(UnresolvedAttribute(name :: Nil), constant: Literal) =>
+                (name, constant.value)
+            }
+            if (!table.partitionColumnNames.exists(resolver(_, attrName))) {
+              throw new AnalysisException(s"${attrName} is not a valid partition column " +
+                s"in table ${table.identifier.quotedString}.")
+            }
+            val dataType = table.partitionSchema.apply(attrName).dataType
+            expr.withNewChildren(Seq(AttributeReference(attrName, dataType)(),
+              Cast(Literal(value.toString), dataType)))
```

Comment:
why do we need to cast a new `Literal`? can't we just use `constant`?

Reply:
constant's dataType may be different with the partition's dataType.
the difference may cause problems for the expression to compare them later.

Reply:
Shouldn't we throw an `AnalysisException` if they have different datatype? I think converting something to string and back to the desired datatype is not a good approach and it may cause issues.
@gatorsmile  what do you think?

Reply:
I think we can't throw AnalysisException for all the situations. e.g.
`CREATE TABLE tbl_x (a INT) PARTITIONED BY (p LONG)`
`ALTER TABLE tbl_x DROP PARTITION (p < 1)`
In this case, the partition's dataType is `LONG` for sure. But the constant's dataType is `INT`

I think it's reasonable to support this situation at least.

Reply:
Ok I get your point. 
I just run a quick test. it threw a exception "java.lang.RuntimeException: Unsupported literal type class org.apache.spark.unsafe.types.UTF8String" at this line when I run:
`ALTER TABLE table_a PARTITION(a < 'test')`
so there is one line change in `literals.scala` needed.
the method `def apply(v: Any): Literal` (literals.scala: line 52) only support `String` not `UTF8String` for now.

Reply:
using the parsed `constant` and if we don't cast it to partition's dataType. it will throw an exception
>java.lang.ClassCastException: java.lang.Integer cannot be cast to java.lang.Long
    at scala.runtime.BoxesRunTime.unboxToLong(BoxesRunTime.java:105)
    at scala.math.Ordering$Long$.compare(Ordering.scala:264)
    at scala.math.Ordering$class.gteq(Ordering.scala:91)
    at scala.math.Ordering$Long$.gteq(Ordering.scala:264)
    at org.apache.spark.sql.catalyst.expressions.GreaterThanOrEqual.nullSafeEval(predicates.scala:710)
    at org.apache.spark.sql.catalyst.expressions.BinaryExpression.eval(Expression.scala:423)

for the case
 `CREATE TABLE tbl_x (a INT) PARTITIONED BY (p LONG)`
`ALTER TABLE tbl_x DROP PARTITION (p >= 1)`
that I mentioned above

Reply:
I see, but what about `Cast(constant, dataType)`?

Reply:
lol. you are right. I will update